openedx/core/djangoapps/course_date_signals/handlers.py

```diff
@@ -0,0 +1,91 @@
+    """
+    result = {}
+    for field_name in date_fields:
+        if field_name not in xblock.fields:
+            continue
+        field = xblock.fields[field_name]
+        if field.scope == Scope.settings and field.is_set_on(xblock):
+            try:
+                result[field.name] = field.read_from(xblock)
+            except TypeError as exception:
+                exception_message = "{message}, Block-location:{location}, Field-name:{field_name}".format(
+                    message=text_type(exception),
+                    location=text_type(xblock.location),
+                    field_name=field.name
+                )
+                raise TypeError(exception_message)
+    return result
+
+
+def extract_dates_from_course(course):
+    """
+    Extract all dates from the supplied course.
+    """
+    log.info('Publishing course dates for %s', course.id)
+    if course.self_paced:
+        metadata = date_field_values(FIELDS_TO_EXTRACT, course)
+        # self-paced courses may accidentally have a course due date
+        metadata.pop('due', None)
+        date_items = [(course.location, metadata)]
+
+        if SelfPacedRelativeDatesConfig.current(course_key=course.id).enabled:
+            duration = get_user_course_duration(user, course)
+            sections = course.get_children()
+            time_per_week = duration/len(sections)
+            due_field = xblock.fields['due']
+            # Apply the same relative due date to all content inside a section,
+            # unless that item already has a relative date set
+            for idx, section in enumerate(sections):
+                items = [section]
+                while items:
+                    next_item = items.pop()
+                    # TODO: Once studio can manually set relative dates,
+                    # we would need to manually check for them here
+                    date_items.append((item.location, {'due': time_per_week*(idx+1)}))
+                    items.extend(item.get_children())
+    else:
+        date_items = []
+        items = modulestore().get_items(course.id)
+        log.info('extracting dates from %d items in %s', len(items), course.id)
+        for item in items:
+            date_items.append((item.location, date_field_values(FIELDS_TO_EXTRACT, item)))
+    return date_items
+
+
+@receiver(SignalHandler.course_published)
+def extract_dates(sender, course_key, **kwargs):  # pylint: disable=unused-argument
+    """
+    Extract dates from blocks when publishing a course.
+    """
+    log.info("Extracting dates from %s", course_key)
+
+    course = modulestore().get_course(course_key)
+
+    if not course:
+        log.info("No course found for key %s", course_key)
+        return
+
+    date_items = extract_dates_from_course(course)
+
+    try:
+        set_dates_for_course(course_key, date_items)
+    except Exception:  # pylint: disable=broad-except
+        log.exception('setting dates for %s', course_key)
```

Comment:
Does that read well? I can't remember how that gets printed. I'm used to seeing a message like "failed to set dates..." but maybe this gets formatted in a way that your phrasing is better.